### PR TITLE
[extractor/iwara] Add access token

### LINKED ISSUE #6671: [Iwara] Unable to access some video without access token
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

Some video is not able to access without 'Authorization' in http request to [https://api.iwara.tv/video/](url)`<video id>`. In the browser, I can login to get this header but yt-dlp isn't able to send it because it's not in the cookie.

The Authorization header string is in the form of 'Bearer ' . concat(e) where e is the accessToken return from [https://api.iwara.tv/user/token](url) request.

~Request~
~https://api.iwara.tv/user/token~
~POST~
~Http header: {`Authorization: Bearer <user token>`}   --> user token was stored in localStorage and accessed by this localStorage.getItem('token')~

~Reponse~
~JSON:~
~accessToken: `<access token>`~

Request
```
https://api.iwara.tv/user/login
POST {
    "email": <username/email>,
    "password": <password>
}
```

Response 
```
JSON: {
    "token": <user token>
}
```

Request
```
https://api.iwara.tv/user/token
POST
Authorization: Bearer <user token>
```

Reponse
```
JSON: {
    "accessToken": <access token>
}
```

Once access token is gained, [https://api.iwara.tv/video/](url)`<video id>` can be request with this header
`Authorization: Bearer <access token>`

Request
```
https://api.iwara.tv/video/<id>
POST
Authorization: Bearer <access token>
```

Reponse
```
JSON: {
    <video info> --> otherwise "message":"errors.notFound" will be returned
}
```

~The problem is I don't know how to get the user token. It seems it was generated from user id and other params and ran through some function to create that that token. I assume this is similar to the X-Version header for download link request.~

The access token can be stored and reused for sub-sequence api requests for other videos.

Seem like they use this for private and login required videos.

List of some video that need this:
token: jlavxivexhegqyge
token: vqrz5sr6j5fegrzv4
404: neglocmyo5tpz7rl4
private: djevysgz9ik044wr
private: 0gqdecjqr1szqv20a

The first 2 will return 404 if Authorization header is not used.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://www.iwara.tv/video/jlavxivexhegqyge']
[debug] User config "C:\Users\Nobita\AppData\Roaming\yt-dlp\config.txt": ['--compat-options', 'filename', '--cookies-from-browser', 'firefox', '--retries', 'infinite', '--retry-sleep', '2', '--extractor-args', 'youtube:skip=dash']
[debug] Encodings: locale cp932, fs utf-8, pref cp932, out cp932 (No VT), error cp932 (No VT), screen cp932 (No VT)
[debug] yt-dlp version stable@2023.03.04 [392389b7d]
[debug] Lazy loading extractors is disabled
[debug] Compatibility options: filename
[debug] Python 3.8.5 (CPython AMD64 32bit) - Windows-7-6.1.7601-SP1 (OpenSSL 1.1.1g  21 Apr 2020)
[debug] exe versions: ffmpeg N-106355-ga24e6ca8bf-20220320 (setts), ffprobe N-106355-ga24e6ca8bf-20220320
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "C:\Users\Nobita\AppData\Roaming\Mozilla\Firefox\Profiles\jdltat92.dev-edition-default-1663626297735\cookies.sqlite"
[Cookies] Extracted 1766 cookies from firefox
[debug] Proxy map: {}
[debug] Loaded 1793 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Available version: stable@2023.03.04, Current version: stable@2023.03.04
yt-dlp is up to date (stable@2023.03.04)
[iwara] Extracting URL: https://www.iwara.tv/video/jlavxivexhegqyge
[iwara] jlavxivexhegqyge: Downloading JSON metadata
ERROR: [iwara] jlavxivexhegqyge: Iwara says: errors.notFound; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "\\192.168.1.40\qemutemp\Temp\yt-dlp\yt_dlp\extractor\common.py", line 694, in extract
    ie_result = self._real_extract(url)
  File "\\192.168.1.40\qemutemp\Temp\yt-dlp\yt_dlp\extractor\iwara.py", line 85, in _real_extract
    raise ExtractorError(f'Iwara says: {errmsg}')


```

Edit

### PR DESCRIPTION
**Pull request by toomyzoom**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Add access token using login info for non private video that is not accessible without account.

Some video would show up as deleted/not exist if not logged in. This is not related to private videos.
If an access token is sent with the http request, the video is accessible. My assumption is those videos are tagged as NSFW/not appropriate for public access.

Fixes #6671

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/iwara.py [hunk 1]
```diff
@@ -79,13 +142,26 @@ def _extract_formats(self, video_id, fileurl):
...
+        username, password = self._get_login_info()
+        if username and password:
+            headers = {
+                'Authorization': f'Bearer {self._get_media_token()}',
+                'Content-Type': 'application/json'
+            }
+        else:
+            headers = {
+                'Content-Type': 'application/json'
+            }
```
Comment (Lesmiscore) [inline yt_dlp/extractor/iwara.py:154]:
```suggestion
        headers = {
            'Authorization': f'Bearer {self._get_media_token()}',
            'Content-Type': 'application/json'
        } if username and password else {  
            'Content-Type': 'application/json'
        }
```

File: yt_dlp/extractor/iwara.py [hunk 2]
```diff
@@ -79,13 +142,26 @@ def _extract_formats(self, video_id, fileurl):
...
         # at this point we can actually get uploaded user info, but do we need it?
         if errmsg == 'errors.privateVideo':
             self.raise_login_required('Private video. Login if you have permissions to watch')
-        elif errmsg:
-            raise ExtractorError(f'Iwara says: {errmsg}')
+        elif errmsg == 'errors.notFound':
+            if not username:
+                self.raise_login_required('Video may need login to view')
+            else:
+                raise ExtractorError('Iwara says: ' + errmsg)
```
Comment (Lesmiscore) [inline yt_dlp/extractor/iwara.py:164]:
```suggestion
        elif errmsg == 'errors.notFound' or not username:
            self.raise_login_required('Video may need login to view')
        else:
            raise ExtractorError(f'Iwara says: {errmsg}')
```

File: yt_dlp/extractor/iwara.py [hunk 3]
```diff
@@ -14,7 +15,49 @@
 )
 
 
-class IwaraIE(InfoExtractor):
+# https://github.com/yt-dlp/yt-dlp/issues/6671
+class IwaraBaseIE(InfoExtractor):
+    _USERTOKEN = None
+    _MEDIATOKEN = None
+    _NETRC_MACHINE = "iwara"
```
Comment (Lesmiscore) [inline yt_dlp/extractor/iwara.py:22]:
```suggestion
    _NETRC_MACHINE = 'iwara'
```

File: yt_dlp/extractor/iwara.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/iwara.py:154]: is `Content-Type` actually needed here? it's not POST

Comment (Lesmiscore) [reply yt_dlp/extractor/iwara.py:154]: Nah not needed lol

**Comment by user1872372839** [comment]
Just a heads up, the problem is still not fixed on nightly 2023.04.24.141204.

Private videos from friends still ask for cookie even though I'm already using them. "Hidden" videos still give a "Iwara says: errors.notFound" error even after loggin in and setting up the cookies.

**Comment by pukkandan** [comment]
> Just a heads up, the problem is still not fixed on nightly 2023.04.24.141204.

hence why https://github.com/yt-dlp/yt-dlp/issues/6671 is still open!

**Review by toomyzoom — COMMENTED** [review]
Added changes follow reviewers suggestion

**Review by Lesmiscore — APPROVED** [review]
I think it's good to merge

File: yt_dlp/extractor/iwara.py [hunk 4]
```diff
@@ -79,12 +142,18 @@ def _extract_formats(self, video_id, fileurl):
...
+        } if username and password else None
+        video_data = self._download_json(f'https://api.iwara.tv/video/{video_id}', video_id, expected_status=lambda x: True, headers=headers)
         errmsg = video_data.get('message')
         # at this point we can actually get uploaded user info, but do we need it?
         if errmsg == 'errors.privateVideo':
             self.raise_login_required('Private video. Login if you have permissions to watch')
-        elif errmsg:
+        elif errmsg == 'errors.notFound' and not username:
+            self.raise_login_required('Video may need login to view')
+        elif errmsg: # None if success
```
[the remark this answers is not in the document]
Comment (toomyzoom) [reply yt_dlp/extractor/iwara.py]: Thanks. Sorry, my first time doing pull request.

**Comment by bobbbob98** [comment]
Anything else need doing with this? I'd like to be able to download these.

**Comment by user1872372839** [comment]
Just an update, ever since today (yesterday it was working perfectly), the login API is not working anymore. If you try to login, all downloads (even those that are not privated/hidden) will fail.

Example: yt-dlp.exe -f b* --username $userhere --password $passwordhere $downloadlink

Result:
[iwara]: Unable to download JSON metadata: HTTP Error 403: Forbidden (caused by <HTTPError 403: 'Forbidden'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

**Comment by bashonly** [comment]
@user1872372839 can you open a new issue about this?

**Pull request merged**